educator/src/Dscp/Educator/Web/Types.hs

```diff
@@ -0,0 +1,79 @@
+-- | Common datatypes for educator and student HTTP APIs
+
+module Dscp.Educator.Web.Types
+       (
+         -- * Flags
+         IsFinal (..)
+       , _IsFinal
+       , HasProof (..)
+
+         -- * Responses
+       , GradeInfo (..)
+       , BlkProofInfo (..)
+
+         -- * Conversions
+       , assignmentTypeRaw
+       ) where
+
+import Control.Lens (Iso', iso, makePrisms)
+import Data.Aeson.Options (defaultOptions)
+import Data.Aeson.TH (deriveJSON)
+import Data.Time.Clock (UTCTime)
+import Servant (FromHttpApiData (..))
+
+import Dscp.Core
+import Dscp.Crypto
+import Dscp.Util.Aeson (AsHex)
+
+-- | Whether assignment is final in course.
+newtype IsFinal = IsFinal { unIsFinal :: Bool }
+    deriving (Eq, Show)
+
+makePrisms ''IsFinal
+
+-- | Whether transaction has been published into public chain.
+newtype HasProof = HasProof { unHasProof :: Bool }
+    deriving (Eq, Show)
+
+data GradeInfo = GradeInfo
+    { giSubmissionHash :: !(Hash Submission)
+    , giGrade          :: !Grade
+    , giTimestamp      :: !UTCTime
+    , giHasProof       :: !Bool
+    } deriving (Show, Eq, Generic)
+
+data BlkProofInfo = BlkProofInfo
+    { bpiMtreeSerialized :: !(AsHex ByteString)
+    , bpiTxs             :: ![PrivateTx]
+    } deriving (Show, Eq, Generic)
+
+---------------------------------------------------------------------------
+-- Simple conversions
+---------------------------------------------------------------------------
+
+assignmentTypeRaw :: Iso' AssignmentType IsFinal
+assignmentTypeRaw = iso forth back
+  where
+    back = \case
+        IsFinal False -> Regular
+        IsFinal True  -> CourseFinal
+    forth = \case
+        Regular     -> IsFinal False
+        CourseFinal -> IsFinal True
```

Comment:
I started loosing track of new `Bool` types we have. Even if `AssignmentType` may become 3-type at some moment.